### PR TITLE
[ONNX] Fix ComputeShapeFromReshape when input_shape_size < reshape_size

### PR DESCRIPTION
**Pull request by jiafatom**
Fixes #{issue number}

### CODE REVIEW

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 1]
```diff
@@ -452,22 +452,22 @@ std::vector<int64_t> ComputeShapeFromReshape(
...
-        shape_ratio *= static_cast<double>(input_shape[i]) / reshape[i];
+        shape_ratio /= static_cast<double>(reshape[i]);
+      } else {
+        shape_ratio /= static_cast<double>(input_shape[i]);
       }
-    } else if (minus_one_pos > -1) {
-      if (i < input_shape_size - reshape_size + minus_one_pos + 1) {
-        shape_ratio *= static_cast<double>(input_shape[i]);
+    } else if (i > minus_one_pos) {
+      if (reshape[i] != 0) {
```
Comment (shubhambhokare1) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:467]: Just curious as to what are cases in which reshape[dim] = 0, Are these cases particular to onnx::reshape nodes from the symbolics?

**Review by shubhambhokare1 — COMMENTED** [review]
Thanks for this @jiafatom 
Overall LGTM, just was curious re, cases where reshape[dim] is zero

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -2281,6 +2281,15 @@ def forward(self, x):
...
 
+    def test_reshape_different_rank(self):
+        class ReshapeModel(torch.nn.Module):
+            def forward(self, x):
+                x = x.reshape(-1, 2, 4, 4, 5, 5)
+                return x
+
+        x = torch.randn(1, 32, 5, 5)
+        self.run_test(ReshapeModel(), x)
+
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:2292]: torch reshape cannot have 0 there.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 1 above)
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:467]: Seems that pytorch exporter does not have reshape[dim]==0, this is for onnx.

Comment (shubhambhokare1) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:467]: As expected. Thanks for confirming.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shubhambhokare1) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:2292]: Maybe you can use something like this
```
def test_reshape_repeat_interleave(self):
        class ReshapeModel(torch.nn.Module):
            def forward(self, x):
                repeats = torch.tensor(4)
                return torch.repeat_interleave(x, repeats, dim=4)

        x = torch.rand(5, 2, 5, 5, 2)
        another_x = torch.rand(15, 2, 10, 10, 2)
        self.run_test(ReshapeModel(), x, test_with_inputs=[another_x],
                      input_names=['input_1'], dynamic_axes={'input_1' : {0 : 'n',
                                                                          2 : 'h',
                                                                          3 : 'w'}})
```

Repeat interleave uses reshape where the reshape input is a combination of 0 (dynamic_axes), -1 (dim for interleaving) and    1 (other dims)

[the remark this answers is not in the document]
Comment (jiafatom) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:2292]: This example indeed has 0 in reshape, but the input shape is not known, see:
```python
  %38 : Long(requires_grad=0, device=cpu) = onnx::Constant[value={-1}]()
  %39 : Long(6, strides=[1], device=cpu) = onnx::Mul(%37, %38)
  %40 : Bool(6, strides=[1], device=cpu) = onnx::Equal(%35, %39)
  %41 : Long(6, strides=[1], device=cpu) = onnx::Where(%40, %37, %35)
  %42 : FloatTensor(device=cpu) = onnx::Expand(%32, %41)
  %43 : Long(5, strides=[1], device=cpu) = onnx::Constant[value= 0  2  0  0 -1 [ CPULongType{5} ]]()
  %44 : Float(*, *, *, *, *, device=cpu) = onnx::Reshape(%42, %43)
```
We cannot get the output tensor value of onnx::Where.

**Comment by jiafatom** [comment]
Based on https://github.com/pytorch/pytorch/pull/55962, I added the shape inference test like:
```python
    def test_reshape(self):
        g = self.create_empty_graph()
        constant = self.insert_tensor_constant(g, torch.ones(2, 16, 5, 5))
        constant_2 = self.insert_tensor_constant(g, torch.tensor([2, 0, -1]))
        shape = g.op("Reshape", constant, constant_2)
        self.run_test(g, shape.node(), expect_tensor('Float', dynamic_shape=(2, 16, 25)))

        g = self.create_empty_graph()
        constant = self.insert_tensor_constant(g, torch.ones(2, 16, 5, 4))
        constant_2 = self.insert_tensor_constant(g, torch.tensor([-1, 0, 4]))
        shape = g.op("Reshape", constant, constant_2)
        self.run_test(g, shape.node(), expect_tensor('Float', dynamic_shape=(10, 16, 4)))

        g = self.create_empty_graph()
        constant = self.insert_tensor_constant(g, torch.ones(2, 16, 5, 4))
        constant_2 = self.insert_tensor_constant(g, torch.tensor([-1, 0, 0]))
        shape = g.op("Reshape", constant, constant_2)
        self.run_test(g, shape.node(), expect_tensor('Float', dynamic_shape=(8, 16, 5)))
```
It passed on my local dev box.
So I get to know how to handle reshape dim = 0 case. There is no index shift.

**Review by BowenBao — APPROVED** [review]
LGTM, thanks David for validating with the new test locally, I can update that test snippet in the PR.

**Review by shubhambhokare1 — APPROVED** [review]
LGTM, thanks for checking for cases with zero :)
